### PR TITLE
Test/run step instructions

### PR DESCRIPTION
**Pull request by Tomi-3-0**
Added full tests for RunStep #117 
- [x] Test `ap += <smth>` is correct.
- [x] Test `jump abs` and `jump rel` jumps are correct.
- [x] Test conditional jumps `jnz` are correct.
- [x] Test `call rel`, `call abs` and `ret` are correct. 

 Bug fix for `error: running instruction: pc update: absolute jump: memory value is not an address`

### CODE REVIEW

File: pkg/vm/vm_test.go [hunk 1]
```diff
@@ -960,6 +971,198 @@ func TestAssertEqualInstruction(t *testing.T) {
...
+	setInitialReg := func(vm *VirtualMachine, regvals ...uint64) {
+		if len(regvals) != 3 {
+			panic("expected three register values")
+		}
+		vm.Context.Ap = regvals[0]
+		vm.Context.Fp = regvals[1]
+		vm.Context.Pc = mem.MemoryAddress{SegmentIndex: 0, Offset: regvals[2]}
+	}
+
+	t.Run("test advancing ap with address", func(t *testing.T){
```
Comment (rodrigo-pino) [inline pkg/vm/vm_test.go:985]:
```suggestion
t.Run("test advancing ap with expression", func(t *testing.T) {
```

File: pkg/vm/vm_test.go [hunk 2]
```diff
@@ -960,6 +971,198 @@ func TestAssertEqualInstruction(t *testing.T) {
...
+
+		writeToDataSegment(vm, vm.Context.Ap, &mem.MemoryAddress{SegmentIndex: 15, Offset: 18})
+
+		err := vm.RunStep(&hintrunner)
+		require.NoError(t, err)
+
+		assert.Equal(t, vm.Context.Pc, mem.MemoryAddress{SegmentIndex: 15, Offset: 18})
+	})
+
+	t.Run("test rel jump with address", func(t *testing.T) {
```
Comment (rodrigo-pino) [inline pkg/vm/vm_test.go:1041]: "test rel jump with expression"

File: pkg/vm/vm_test.go [hunk 3]
```diff
@@ -960,6 +971,198 @@ func TestAssertEqualInstruction(t *testing.T) {
...
+func TestCallAndRetInstructions(t *testing.T){
+	hintrunner := noHintRunner{}
+	setInitialReg := func(vm *VirtualMachine, regvals ...uint64) {
+		if len(regvals) != 3 {
+			panic("expected three register values")
+		}
+		vm.Context.Ap = regvals[0]
+		vm.Context.Fp = regvals[1]
+		vm.Context.Pc = mem.MemoryAddress{SegmentIndex: 0, Offset: regvals[2]}
+	}
```
Comment (rodrigo-pino) [inline pkg/vm/vm_test.go:1105]: You have the same function repeated across all tests. I think is good to only define it once

**Review by rodrigo-pino — COMMENTED** [review]
Great work! Minor comments and it is good to go

File: pkg/vm/vm_test.go (hunk 3 above)
Comment (Tomi-3-0) [reply pkg/vm/vm_test.go:1105]: Noted

File: pkg/vm/vm_test.go [hunk 4]
```diff
@@ -862,7 +873,7 @@ func (r *noHintRunner) RunHint(_ *VirtualMachine) error {
 	return nil
 }
 
-func TestAssertEqualInstruction(t *testing.T) {
+func RunAssertEqualInstruction(t *testing.T) {
```
Comment (rodrigo-pino) [inline pkg/vm/vm_test.go:876]: Why this tests are here?

File: pkg/vm/vm_test.go [hunk 5]
```diff
@@ -960,6 +974,173 @@ func TestAssertEqualInstruction(t *testing.T) {
 
 }
 
+func TestRunStepInstructions (t *testing.T){
```
Comment (rodrigo-pino) [inline pkg/vm/vm_test.go:977]: And not here?

Comment (Tomi-3-0) [reply pkg/vm/vm_test.go:977]: okay

File: pkg/vm/vm_test.go [hunk 6]
```diff
@@ -958,6 +972,160 @@ func TestAssertEqualInstruction(t *testing.T) {
...
+
+		writeToDataSegment(vm, vm.Context.Ap+4, 5)
+
+		err := vm.RunStep(&hintrunner)
+		require.NoError(t, err)
+
+		// The ret instruction resets the value of fp
+		// ap is advanced by 2, and fp is set to the new ap.
+		assert.Equal(t, vm.Context.Fp, vm.Context.Ap, uint64(7))
+	})
```
Comment (rodrigo-pino) [inline pkg/vm/vm_test.go:1128]: You are not testing `ret` here, you are retesting `call rel [ap + 4]`

**Review by rodrigo-pino — APPROVED** [review]
Great!

**Pull request merged**